### ISSUE TITLE
PostGIS support

### ISSUE THREAD

**Issue by steebchen** [prisma/prisma#2789]
[PostGIS](postgis.net) is a popular extension to support geolocation data types and queries in Postgres. We should think about how we can add support for PostGIS, both for data types and how querying data would look like.

**Comment by ShanonJackson** [comment]
Hope coming?

**Comment by reslear** [comment]
is planned before 2022?

**Comment by AviSantoso** [comment]
+1

**Comment by garytube** [comment]
Same. It would be helpfule if Prisma Migrate would save the the extensions when resetting the database.

`CREATE EXTENSION postgis;`   needs to be added to the migration_xxx.sql file


```
Error: Database error
Error querying the database: db error: ERROR: type "geography" does not exist
   0: migration_core::api::ApplyMigrations
             at migration-engine/core/src/api.rs:73
```

If you create the extension by hand before running prisma migrate you can prevent this error

**Comment by reslear** [comment]
@garytube of course, @matthewmueller @pantharshit00 yes need support most popular types `geography` and `geometry`)

**Comment by ShanonJackson** [comment]
@pantharshit00 ? 

Can we get some kind of response; I feel like there is overwhelming support for this; And personally this is blocking us from using Prisma on a few other projects running PostGiS

**Comment by VinayaSathyanarayana** [comment]
When is this expected?

**Comment by psmaAaron** [comment]
A big +1 for this!

I know for my use cases, having the resulting geometry in GeoJSON is perfect. A simple wrapping of the Geometry columns in `ST_AsGeoJSON` would be a great start. It can then be fed into [Turf](https://github.com/Turfjs/turf) for processing and/or displayed in [Leaflet](https://github.com/Leaflet/Leaflet)/[MapLibre GL JS](https://github.com/maplibre/maplibre-gl-js)/whatever quite easily!

**Comment by MurzNN** [comment]
As first little starting step will be good to add support for `Point` postgres field type, because it is very popular in projects even without PostGIS extension.

**Comment by j0hannr** [comment]
+1

**Comment by gmcquistin** [comment]
+1 , this is a showstopper for migrating geospatial projects from eg. Ruby/Rails to Node

**Comment by kjkurtz** [comment]
+1 to this. This is a showstopper for us and has resulted in us continuing to use Knex with the Knex-Postgis extension. Almost all of our data has a geometry column due to the nature of land construction work.

**Comment by ghyath5** [comment]
Is there any updates here 👀

**Comment by boredland** [comment]
I started migrating to mikro-orm bc there adding custom types is possible.

**Comment by s875515** [comment]
+1

**Comment by masihjahangiri** [comment]
It's very important for us.

**Comment by DanielCavalcante** [comment]
+1

**Comment by ovidb** [comment]
Has anyone tried moving the needle on this? Is there anything we can do to get this moving?

**Comment by egluhbegovic** [comment]
What is the consensus on this one?
It's a make-or-break decision for me. 

If geometry queries are not supported I will have to go with knex.

Is there an ETA?

**Comment by kjkurtz** [comment]
@egluhbegovic I ended up going with knex for this very reason.

**Comment by janpio** [comment]
Right now it is not supported and even if we would start work on it right now, it would take a few weeks to be finished and released. Sorry about that. We hope we will be able to win you back when we release support for this.

**Comment by boredland** [comment]
> What is the consensus on this one?
> It's a make-or-break decision for me. 
> 
> If geometry queries are not supported I will have to go with knex.
> 
> Is there an ETA?

I can only recommend Mikro-ORM (it works on top of knex). You can defined your custom types with custom methods and that is a really nice.

**Comment by RafaelKr** [comment]
There is a workaround for PostGIS, so you can use Prisma for all queries which DON'T use PostGIS specific types. But for every query which requires a PostGIS type you need to do a raw query with `pg` instead of the prisma client due to [this issue](https://github.com/prisma/prisma/issues/2847).  
For the database I used a `kartoza/postgis` docker container. You can have a look at [the reproduction Repo](https://github.com/RafaelKr/prisma-repro-2847) I created for that issue. I didn't test it with the latest Prisma but it should still be working.

Also I used this approach in a production project and there I added the pg client to the context, so I can use `ctx.pg.query`:

<details>
<summary>src/context.ts</summary>

```ts
import { PrismaClient } from '@prisma/client'
import { Pool } from 'pg'

const prisma = new PrismaClient()

// TODO: pg can be replaced with prisma client raw queries when this issue is resolved:
// https://github.com/prisma/prisma/issues/2847
const pg = new Pool({
	connectionString: process.env.POSTGRES_URL,
})

export interface Context {
	prisma: PrismaClient
	pg: Pool
}

export function createContext(): Context {
	return {
		prisma,
		pg,
	}
}

```

</details>

<details>
<summary>src/schema/utils/locations-within-area.helper.ts (NOT the top-level schema folder)</summary>

```ts
import { Context } from '~/context'
import { AreaCenter } from '~/schema/models/AreaCenterInput'

interface QueryResultRow {
	id: string
}

interface QueryResult {
	rows: QueryResultRow[]
}

const query = `
SELECT
	id
FROM
	impegio_prisma2."JobLocation"
WHERE
	ST_DWithin(ST_MakePoint(lng, lat)::geography, ST_MakePoint($1, $2)::geography, $3)
`

export const queryLocationsWithinAreaIds = async (areaCenter: AreaCenter, ctx: Context) => {
	// TODO: pg can be replaced with prisma client raw queries when this issue is resolved:
    // https://github.com/prisma/prisma/issues/2847
	const jobLocationsWithinArea: QueryResult = await ctx.pg.query(query, [areaCenter.lng, areaCenter.lat, areaCenter.radius])
	return jobLocationsWithinArea.rows.map((location) => location.id)
}

```

</details>

Hope this can help somebody.

**Comment by sunech** [comment]
I would be very curious about an ETA as well, as it could be kind of a deal breaker for our need.

**Comment by boredland** [comment]
> But for every query which requires a PostGIS type you need to do a raw query with pg

Yeah... But why didn't you do the raw query with the raw query client that is already integrated into Prisma or am I missing something. At least that is what I did before switching to Mikro-ORM.

**Comment by egluhbegovic** [comment]
@boredland IF you were already using Prisma, and were able to find a way to manage geo queries.
Why migrate to Miko-ORM?

**Comment by RafaelKr** [comment]
> > But for every query which requires a PostGIS type you need to do a raw query with pg
> 
> Yeah... But why didn't you do the raw query with the raw query client that is already integrated into Prisma or am I missing something. At least that is what I did before switching to Mikro-ORM.

Maybe it's also working with the latest Prisma now. 2 years ago I had [this issue](https://github.com/prisma/prisma/issues/2847) (which I also mentioned in my comment) and the issue was never closed as fixed. I didn't test it with the latest Prisma version, though.

**Comment by boredland** [comment]
> @boredland IF you were already using Prisma, and were able to find a way to manage geo queries. Why migrate to Miko-ORM?

because raw queries are stupid to use and defy every reason I used prisma (type safety).

> @boredland Maybe it's also working with the latest Prisma now. 2 years ago I had [this issue](https://github.com/prisma/prisma/issues/2847) (which I also mentioned in my comment) and the issue was never closed as fixed. I didn't test it with the latest Prisma version, though.

I did quit prisma before I got too deep into using raw queries, but iirc I used st_fromgeojson and st_togeojson, why the last comment in that issue indicates people using st_buffer etc. - perhaps I was lucky and never reached the limits?

**Comment by garretteklof** [comment]
Yeah, we use raw queries for everything on the Prisma client for geospatial queries. It's a huge pain in the you-know-what, and unfortunately we are in too deep now. It was obviously a bad play .. I just figured an OS project with funding would be far more efficient at pushing out features on their roadmap. And I don't mean that disrespectfully, but at the same time, ladies and gents. I'm not wrong. It was said on a few comments that the feature would be pushed out by the end of 2021 at the latest.

**Comment by j0hannr** [comment]
@garretteklof Prisma updated their [Roadmap](https://www.notion.so/Prisma-Roadmap-50766227b779464ab98899accb98295f) and **PostGIS/GIS support** is now planned. It would be nice to know a *rough* delievery date @matthewmueller :)

**Comment by garretteklof** [comment]
@j0hannr thank you. I believe it was 'Planned' before (I remember @janpio at least confirming at some point) .. I would love to see it scooted over to `Early Availability` at some point in the near-ish future 😉

**Comment by janpio** [comment]
> It was said on a few comments that the feature would be pushed out by the end of 2021 at the latest.

I doubt that, as we generally do not make ETA statements about features before they are done -  because those always go sideways and then you need to pedal back and apologize (I worked in game development some years ago, so I know what I am talking about here 😆).

So we are definitely aware that GIS support is well requested and needed and are obviously working on adding it to Prisma. We would love if that could happen sooner than later, but again, do not have an ETA right now.

**Comment by garretteklof** [comment]
Haha touché @janpio .. I would've sworn by it too; alas, I scoured the Github issues and could not find any verifiable proof of this alleged confirmation. I must've dreamt it late one night in a fit of geospatial reverie. You win this time, sir 😜

**Comment by kamwing** [comment]
+1

**Comment by ngprnk** [comment]
Support for postgis would be a dealbreaker for us, right now we are using knex with postgis.

**Comment by olafurr** [comment]
+1

**Comment by KinjalWS** [comment]
+1

**Comment by fuergaosi233** [comment]
+1

**Comment by revmischa** [comment]
For inspiration, maybe consider the very popular GeoAlchemy extension for SQLAlchemy - https://geoalchemy-2.readthedocs.io/en/latest/orm_tutorial.html
They define geography and geometry data types and have a namespace of functions that are present in PostGIS
So you can do stuff like `query = session.query(Lake).filter(Lake.geom.ST_Contains('POINT(4 1)'))`

**Comment by campbellfleury** [comment]
Was interesting to see during the Prisma Day presentations that PostgreSQL extension support is in the roadmap. Anywhere we can see ideations or mock-ups of how this might work? Similar to how they showed schema support.

**Comment by matthewmueller** [comment]
Thanks for the inspiration @revmischa!

@campbellfleury no mockups yet, but we'll post a design in this issue once we have some more ideas of how that might look.

**Comment by davecthomas** [comment]
For now, for those of us using basic GIS features in our apps, I guess we just store lat, lng as point data type in Postgres? This will leave lots of computation requirement on the app developer.

**Comment by matthewmueller** [comment]
You can also define and manage PostGIS indices with Prisma migrate in [v3.14](https://github.com/prisma/prisma/releases/tag/3.14.0).

That in combo with the raw query improvements (also in 3.14.0) should hopefully give you a somewhat reasonable workaround for those PostGIS-specific queries until we have better native support.

**Comment by floelhoeffel** [comment]
Hello everyone 👋

We are kicking off design of GIS support for Prisma!

If you would like to help, please let us learn about your GIS needs!

**Via survey:**
Please fill out our [GIS User Research Survey](https://prisma103696.typeform.com/to/p8xo5o95) 

**Via email:**
Alternatively, just shoot us an an email to [feedback@prisma.io](mailto:feedback@prisma.io).
The questions we have are (all in the context of GIS):
- what is your GIS use case?
- what database do you use?
- what column types do you use?
- what queries do you use?
- what indexes do you use?
- what kind of defaults do you use?

Thank you!

**Comment by yasha-spare** [comment]
@floelhoeffel just checking in, any update on Prisma GIS support?  I'm a big fan of Prisma, but for projects that make significant use of PostGIS, the current hacks are quite nasty, to the point where Prisma becomes too much of a pain to use, and I have to use a different ORM.

Also, if there are designs and an in-progress feature branch, is this something where you'd be looking for community contributors?

**Comment by teddarcuri** [comment]
I know there is a ton of community support for this request, but I just want to throw one more +1 in on this.
Been a real bummer working around this. This feature will be legit cause for celebration 🎉

**Comment by floelhoeffel** [comment]
Thanks so much for all the input - keep it coming! Especially the survey has helped us greatly understand a good scope for this feature. We are planning the next cycle first week of December and will provide an update shortly afterwards.

Thanks so much for the patience 🙇‍♂️

**Comment by segheysens** [comment]
@floelhoeffel Hi there! Any timeline updates? I'm building a serverless app, so I've added some API endpoints that just query the database directly, but being able to write the queries with Prisma (like most of my other resource routes) would be great.

**Comment by MurzNN** [comment]
The first good steep can be adding a simple "point" data type (lat, lon), which is supported by many database engines.

**Comment by stevokk** [comment]
We use the common queryRaw workaround in our endpoints (inc a very limited set of filters and sorts as a patch until postgis support lands)

One thing we found this morning is that prisma Params.args.modelName doesn't populate even if we add the typing on the returning list as we can't call queryRaw on the model, but only the prisma client. Luckily it only caused cache misses here but an easier escape hatch is needed if postgis is still in the design stage. Allowing model.queryRaw would suit - if the developer doesn't return the right types it is acceptable to fail?

Love the library apart from these small issues!

**Comment by dagual** [comment]
Geographical queries are crucial for our project, we need the simplest functionality: just find the nearest object on the map by coordinates. But since this is not available in Prisma, our team has to use another framework to work with them, we will be very glad to see this feature implemented in Prisma!

**Comment by KinjalWS** [comment]
> Geographical queries are crucial for our project, we need the simplest functionality: just find the nearest object on the map by coordinates. But since this is not available in Prisma, our team has to use another framework to work with them, we will be very glad to see this feature implemented in Prisma!

Hey @dagual!
May I know what other framework(s) you and your team uses to handle Geographical queries?

**Comment by KiwiKilian** [comment]
@KinjalWS [TypeORM does support spatial](https://github.com/typeorm/typeorm/blob/master/docs/entities.md#spatial-columns) columns and we successfully use it with PostGIS.

**Comment by KinjalWS** [comment]
> @KinjalWS [TypeORM does support spatial](https://github.com/typeorm/typeorm/blob/master/docs/entities.md#spatial-columns) columns and we successfully use it with PostGIS.

I was wondering how using multiple ORMs in a single Application server and single database will affect? I'm concerned most about how to deal with migrations across the different ORMs. Like if I create a table in TypeORM and migrate through same, wont it affect the drift in the PrismaORM's migration history?

**Comment by dagual** [comment]
> 
Hi, we use drupal's build in search api, to search for nearest geo objects by coordinates.  Hope, when geo queries will be available in prisma we will be able to switch from legacy part of project.

**Comment by dagual** [comment]
> Hey @dagual!
> May I know what other framework(s) you and your team uses to handle Geographical queries?

Hi, we use drupal's build in search api, to search for nearest geo objects by coordinates. Hope, when geo queries will be available in prisma we will be able to switch from legacy part of project.

**Comment by grinono** [comment]
> rst week of December and wil

Any update on this, we waited to implement the raw query for Prisma to launch this beta feature in December. Can we better not wait any longer?

**Comment by gwro** [comment]
Hey, any update? It's crucial for our project as well. :)

**Comment by grinono** [comment]
We implemented it from the raw query.

Prisma schema >
``` model Delivery {
  id          String                               @id @default(uuid())
  createdAt   DateTime                             @default(now()) @map("created_at")
  createdBy   String?                              @map("created_by")
  user        User?                                @relation(fields: [createdBy], references: [id])
  longitude   String
  latitude    String
  coordinates Unsupported("geometry(Point, 4326)")
  order       Order[]

  @@index([coordinates], name: "location_idx", type: Gist)
  @@map("delivery")
}
```


input query >

```
    await prismaClient.$queryRawUnsafe(`
    INSERT INTO delivery (id, coordinates, created_by, latitude, longitude)
    VALUES ('${id}',ST_GeomFromText('POINT(${items.latitude} ${items.longitude})', 4326), '${items.createdBy}', '${items.latitude}' , '${items.longitude}');
    `)
```

**Comment by tonymckendry** [comment]
@grinono - I am pretty sure Longitude comes before Latitude in your POINT setup. Thanks for the example!

**Comment by brianbolnick** [comment]
Following! Hoping to see something come to life here soon 🙌

**Comment by benjichat** [comment]
+1

**Comment by darcy-vitacca** [comment]
+1

**Comment by lh0x00** [comment]
+1

**Comment by MurzNN** [comment]
I've found a great article with describing how to use PostGIS with Prisma already now: https://freddydumont.com/blog/prisma-postgis

**Comment by darcy-vitacca** [comment]
> I've found a great article with describing how to use PostGIS with Prisma already now: https://freddydumont.com/blog/prisma-postgis

This is great, have you got anywhere where I could learn about you to query this from the db given a lat long?

**Comment by MurzNN** [comment]
> > I've found a great article with describing how to use PostGIS with Prisma already now: https://freddydumont.com/blog/prisma-postgis
> 
> This is great, have you got anywhere where I could learn about you to query this from the db given a lat long?

Something like this mb?
https://towardsdatascience.com/spatial-data-science-with-postgresql-geometries-c00387755700
https://forrest.nyc/spatial-sql-cookbook/

**Comment by tahaontech** [comment]
I suggest a bigger solution to this problem [in this issue](https://github.com/prisma/prisma/issues/18281)   
I know there are a few other extensions just for PostgreSQL like PgRouting and a lot of custom data types,   
but we can use them all with my solution without waiting for prisma to support them and **without complete row query**
and that solution works for other databases like SQL server at the same time. :)

**Comment by MurzNN** [comment]
https://github.com/prisma/prisma/issues/2789#issuecomment-1405221231

@grinono, did you invent some tricks to get the 'coordinates' field value on Prisma SELECT queries, without doing a separate `$rawQuery` to get the value?

**Comment by grinono** [comment]
no, we just use the queryRawUnsafe, waiting for prisma to include a better query method to query custom fields or specific postgis fields.

**Comment by MurzNN** [comment]
I've filled a feature request to add support casting to string (or any other supported format) unsupported fields: https://github.com/prisma/prisma/issues/18544 that, I hope, should help with retrieving data from GIS raws in Prisma queries.

**Comment by MikeyUchiha** [comment]
@janpio 3 years later and the Prisma team still hasn't figured out how to add this? Can you please get this feature on the roadmap? This makes many apps completely unable to use Prisma and have to explore other options because your PostGIS is so terrible. Every other major ORM now supports this and you're getting left behind.

**Comment by lewebsimple** [comment]
@MikeyUchiha with the recently released [Prisma Client extensions](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions) it's now easier to add fields / queries with custom logic such as PostGIS polygon, etc.

... or you can use another ORM if they're so great.

**Comment by FelixZY** [comment]
There were some official responses regarding PostGIS on reddit the other day: https://www.reddit.com/r/node/comments/14954z9/comment/jo4jzyh/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

Summary:
* Official word is: "For now, the best way to interact with PostGIS is by using `$queryRaw`"
* PostGIS seems prioritized but not started (https://prismaio.notion.site/Prisma-ORM-Roadmap-50766227b779464ab98899accb98295f)
* [SafeQL](https://safeql.dev/compatibility/prisma.html) was suggested to help with raw query linting. Not personally sure how this interacts with prisma's `Unsupported(...)` fields though.
* The dev at https://tryjourneyapp.com/ was going to try Prisma Client Extensions out and possibly open source his solution

**Comment by acao** [comment]
I recently worked on a project where I went through four different typescript ORMs trying to find one thst worked nicely with PostGIS. I may write a blog post about this, with some useful tips for ORM authors who want to provide better PostGIS. Another thing to consider, especially if your service is still greenfield, is whether ORM architecture is best, or a custom DBAL using the drivers is the best fit for your software design.

**Comment by jeffschuler** [comment]
Write that blog post Rikki! 👋🏼

**Comment by acao** [comment]
wow, @jeffschuler! it's been a while!! 😂

**Comment by acao** [comment]
Here is a rough draft of the article! It's focused on what I imagine are the most common use cases for my fellow PostGIS users. I tried to do my best to rebuild all the domain context I built up a few months ago, without being able to see any of the code I wrote!

https://github.com/acao/rikki.dev/blob/d82327a8dd3ab8437840f9bbf1935fb3c2d657d2/pages/posts/postgis-typescript-evaluation.md

Please, if you have any critique or need more advice on other dimensions of this wonderful Postgres extension, feedback in the PR comments is very welcome!

https://github.com/acao/rikki.dev/pull/1

**Comment by MikeyUchiha** [comment]
thank you @acao  This is really helpful. I think your analysis of Supabase was pretty biased though. You can create a custom function using SQL with the exact WHERE clause that you need and then expose it. https://supabase.com/docs/guides/database/extensions/postgis

I'm thinking of just going with Drizzle though since it's much faster and if I have to write SQL anyway, I might as well get a performance boost. I really wanted to use Prsima for the ease of use and Studio editor but based on how long it's taken them to implement something so common and simple, I don't think it's worth the wait.

**Comment by acao** [comment]
@MikeyUchiha the problem we had with this was the custom sql functions were hard to manage and version. you have to write migrations for every change to the sql functions 😭

**Comment by MikeyUchiha** [comment]
If you make the function dynamic, not sure why you'd need to make more than one change.

**Comment by acao** [comment]
this was the issue supabase was never able to resolve for us :(

https://github.com/orgs/supabase/discussions/14013

**Comment by MikeyUchiha** [comment]
Yikes! Sorry to hear that. So far Drizzle team has been responsive at least. Maybe check it out. :)

**Comment by acao** [comment]
@MikeyUchiha oh yes, supabase was very responsive in support channels, but they just didn't seem to be able to make progress on the issue.  The client is very happy with typeorm and their hosted database provider, they didn't need anything more than a managed database provider, but drizzle looks cool!

**Comment by jointhejourney** [comment]
Thanks for the mention @FelixZY, I'm still trying to figure out the best way to go about creating an extension that is easy to set up, but for now this is how I am doing with prisma result extensions:

```ts
// set a function fetchLocation() on any results for the model "Feature"
const xprisma = prisma.$extends({
  result: {
    feature: {
      fetchLocation: {
        needs: { id: true },
        compute(data) {
          // a model Feature has a GEOGRAPHY field "location", which we fetch as a geojson using ST_AsGeoJSON
          return async () => {
            const locationRaw = await prisma.$queryRaw<
              { location: string }[]
            >`SELECT ST_AsGeoJSON(location) as location FROM public."Feature" WHERE id::text = ${data.id} LIMIT 1`;
            const location = JSON.parse(locationRaw[0].location) as Point;
            return location;
          };
        },
      },
    },
  },
});

const feature = await xprisma.feature.findUnique({
  where: {
    id: '06ef604e-4ab8-4497-aa13-80e3f84061d7',
  },
});

// then you can fetch the location for each result
const location = feature?.fetchLocation();
```

This is not ideal since you'd have to do N calls for each item if you're doing a fetch many. I've been looking into modifying the queries themselves during runtime, and I'll post here if I figure something better.

**Comment by FelixZY** [comment]
Thank you for sharing @jointhejourney . Based on your example, I was able to produce what I needed. Sharing my implementation here for others who may be interested: https://github.com/dansdata-se/api/commit/fd19131ccb5ed3bec81f1b44281f836ef7edae6d#diff-eb4535afb8c75433ea44c91e28022cd4949a9ab06b2cd6ce9848dca61ff8611d

**Comment by guyromb** [comment]
any update?

**Comment by dxm1244** [comment]
Any updates?

**Comment by dxm1244** [comment]
Any updates?

**Comment by markymc** [comment]
@janpio some kind of statement on this would be nice for the Prisma community.

**Comment by lh0x00** [comment]
Back here after a long time, and everything is still the same as before. I'm not sure, but I have a short break this weekend, I'll try working on this issue to see how it goes.
If it's positive, I'll continue to complete it. Hope to receive help when needed.

**Comment by janpio** [comment]
I can't give any other statement than that we are aware of the feature request, that we very much like the feature request, but right now can not prioritize working on it as we are busy with other things. I hope that will change soon, and if it does I will certainly let you know.

**Comment by holy-dev** [comment]
> > Hello everyone 👋
> > We are kicking off design of GIS support for Prisma!
> 
> Would love to have even just nearby search (points within a radius from lat/lng) and bbox search (points within a bbox) for a v0.1. That would probably satisfy 90% of the use case for everyone.

You're correct. Most of us need nearby search only.

**Comment by hxtruong6** [comment]
Are there any update from PostGIS with Prisma? Or any workaround way?

**Comment by iamclaytonray** [comment]
My workaround for this was to just index specific data into Typesense and use that as my "PostGIS" data layer. I don't store too much data there. Because I also need to search/filter alongside querying within geo bounds, I added all of the necessary fields to filter by and then a `profileId`. From there, I get the response from Typesense and query my database for those returned `profileId`s.

Sub-optimal but does the job

**Comment by martinratinaud** [comment]
Hi all, ss there an ETA for this? (I wonder because issue was opened 4 years ago)

Thanks

**Comment by corwinstephen** [comment]
Need this

**Comment by hansinator** [comment]
+1

**Comment by leonlee723** [comment]
> Here's my current solution:
> 
> 
> 
> Schema (make sure to add the postgis extension to the top of the migration (`CREATE EXTENSION IF NOT EXISTS postgis;`)
> 
> ```prisma
> 
> model UserLocation {
> 
>   id       String                               @id @default(dbgenerated("uuid_generate_v4()")) @db.Uuid
> 
>   location Unsupported("geography(Point,4326)")
> 
>   user     User                                 @relation(fields: [userId], references: [id])
> 
>   userId   String                               @unique @db.Uuid
> 
> 
> 
>   @@map("user_locations")
> 
> }
> 
> 
> 
> model User {
> 
>   location             UserLocation?
> 
>   
> 
>   @@map("users")
> 
> }
> 
> ```
> 
> 
> 
> Helper methods specific to my use-case, but they are pretty generic. If you need help, Cursor (IDE) is great or any LLM.
> 
> ```ts
> 
> import { prisma } from '../services/prisma.js';
> 
> 
> 
> export const createUserLocation = async (
> 
>   userId: string,
> 
>   latitude: number,
> 
>   longitude: number,
> 
> ) => {
> 
>   const point = `POINT(${longitude} ${latitude})`;
> 
>   await prisma.$queryRaw`
> 
>         INSERT INTO "user_locations" ("userId", location) 
> 
>         VALUES (${userId}::uuid, ST_GeomFromText(${point}, 4326))`;
> 
> };
> 
> 
> 
> export const updateUserLocation = async (
> 
>   userId: string,
> 
>   latitude: number,
> 
>   longitude: number,
> 
> ) => {
> 
>   const point = `POINT(${longitude} ${latitude})`;
> 
>   await prisma.$queryRaw`
> 
>         UPDATE "user_locations"
> 
>         SET location = ST_GeomFromText(${point}, 4326)
> 
>         WHERE "userId" = ${userId}::uuid`;
> 
> };
> 
> 
> 
> interface UserLocationRaw {
> 
>   userId: string;
> 
>   location: string;
> 
>   distance_miles: number;
> 
> }
> 
> 
> 
> export const getNearbyUsers = async (
> 
>   latitude: number,
> 
>   longitude: number,
> 
>   radius: number,
> 
> ): Promise<UserLocationRaw[]> => {
> 
>   const point = `POINT(${longitude} ${latitude})`;
> 
> 
> 
>   const formattedRadius = radius * 1609.344;
> 
> 
> 
>   // Find all user locations within a 10-mile radius
> 
>   const nearbyLocations = await prisma.$queryRaw`
> 
>       SELECT "userId", ST_AsText(location) as location,
> 
>         ST_Distance(
> 
>           location::geography,
> 
>           ST_GeomFromText(${point}, 4326)::geography
> 
>         ) / 1609.344 as distance_miles
> 
>       FROM "user_locations"
> 
>       WHERE ST_DWithin(
> 
>         location::geography,
> 
>         ST_GeomFromText(${point}, 4326)::geography,
> 
>         ${formattedRadius}
> 
>       )`;
> 
> 
> 
>   return nearbyLocations as UserLocationRaw[];
> 
> };
> 
> 
> 
> interface UserLocation {
> 
>   userId: string;
> 
>   location: string;
> 
>   longitude: number;
> 
>   latitude: number;
> 
> }
> 
> 
> 
> interface UserLocationReturn {
> 
>   userId: string;
> 
>   longitude: number;
> 
>   latitude: number;
> 
> }
> 
> 
> 
> export const getUserLocation = async (
> 
>   userId: string,
> 
> ): Promise<UserLocationReturn> => {
> 
>   const userLocation: UserLocation[] = await prisma.$queryRaw`
> 
>   SELECT "userId", ST_AsText(location) as location,
> 
>          ST_X(location::geometry) as longitude,
> 
>          ST_Y(location::geometry) as latitude
> 
>   FROM "user_locations"
> 
>   WHERE "userId" = ${userId}::uuid`;
> 
> 
> 
>   if (userLocation && userLocation[0]) {
> 
>     return {
> 
>       userId: userLocation[0].userId,
> 
>       longitude: parseFloat(userLocation[0].longitude.toString()),
> 
>       latitude: parseFloat(userLocation[0].latitude.toString()),
> 
>     };
> 
>   }
> 
> 
> 
>   return userLocation as unknown as UserLocationReturn;
> 
> };
> 
> 
> 
> ```
> 
> 
> 
> Example implementation: (I've broken this down to be pretty contrived but you should get the idea)
> 
> ```ts
> 
> const main = async () => {
> 
>   const nearbyUsers = await getNearbyUsers(latitude, longitude, maxRange);
> 
> 
> 
>   const profiles = await prisma.profile.findMany({
> 
>       where: {
> 
>           user: {
> 
>             id: {
> 
>               in: nearbyUsers.map((user) => user.userId),
> 
>             },
> 
>         },
> 
>       },
> 
>     });
> 
> }
> 
> ```
> 
> 
> 
> ---
> 
> 
> 
> Obviously, actual support would be ideal but hopefully this helps anyone that needs some help. It's definitely not the most efficient (database level or dev experience) but it does get the job done. 
> 
> 
> 
> I haven't played around with the typed sql or sql gen feature that prisma released not too long ago but that could possibly make this code a bit better too. 
> 
> 
> 
> And last bit of context:
> 
> 
> 
> I am running this in an app that has over 30 tables and a TON of joins. My actual schema is fairly complex, and some of my queries are pretty complex. I haven't really seen much degradation in perf but that's likely because the one geo query pales in comparison in going 7 tables deep to get relational data. That said, I still have good performance. On the other side of the spectrum, I do think this actually works pretty well. If you come up with something similar, I'd be curious to see that code (or a contrived example). 

Well done.

**Comment by spacerrsdreams** [comment]
+1

**Comment by PetyaOgurkin** [comment]
+1

**Comment by stephane-segning** [comment]
+1

**Comment by zeayal** [comment]
+1

**Comment by SAE-Esferico2** [comment]
+ 1

**Comment by spacerrsdreams** [comment]
PLEASEEEEEEE LETS ADD THIS <3

**Comment by JCluzet** [comment]
+1

**Comment by ofekBerko** [comment]
+1

**Comment by amyewang** [comment]
+1!

**Comment by spacerrsdreams** [comment]
+1

**Comment by LeonardoDB** [comment]
+1

**Comment by jamalahmed1122-teamx** [comment]
+1

**Comment by jthomaschewski** [comment]
Please avoid posting “+1” comments. Instead, add a 👍 reaction to the original post.
This helps keep the thread clean and prevents unnecessary notifications for everyone following the issue. Thanks!

**Comment by iamclaytonray** [comment]
@jthomaschewski I would normally agree but this is such fundamental database behavior, I think Prisma team should borderline be spammed on this issue. The silence here is deafening.

**Comment by ian** [comment]
Anyone interested in pooling together for a git bounty? I've wanted this feature for years but I don't have time to contribute myself. Happy to throw some $$$ to the cause.

**Comment by jannis6023** [comment]
How much money would be required to realize the implementation in reasonable time?Am 25.08.2025 um 20:27 schrieb ian ***@***.***>:﻿ian left a comment (prisma/prisma#2789)
Anyone interested in pooling together for a git bounty? I've wanted this feature for years but I don't have time to contribute myself. Happy to throw some $$$ to the cause.

**Comment by ian** [comment]
@jannis6023 I've seen git bounties in the $700-$1000 get snatched up quickly.

**Comment by marcospgp** [comment]
is there an update on postgis support in prisma?

**Comment by noelmcloughlin** [comment]
GIS is big business (Gov, UN, EU, etc) so needed.  I'm looking at [this](https://freddydumont.com/blog/prisma-postgis), [this](https://blog.damianesteban.dev/blog/how-to-use-postgis-with-prisma/) and https://github.com/prisma/prisma/issues/2789#issuecomment-2477862028 to get ideas.

**Comment by hanjpk** [comment]
Please... is it still not available?

**Comment by ahmadjoya** [comment]
any update?

**Comment by AdrianPaulV** [comment]
Support for PostGIS please!!!

**Comment by samresh9** [comment]
Support for PostGIS please

**Comment by oskolsky** [comment]
Support for PostGIS please

**Comment by SentryAryan** [comment]
Support for PostGIS please

**Comment by lh0x00** [comment]
Ref https://github.com/prisma/prisma/pull/29365 https://github.com/prisma/prisma-engines/pull/5797

**Comment by lh0x00** [comment]
I’ve created a PR for this feature. I mentioned it before, but I only just found the time to work on it today. If you’re happy with it, feel free to leave a reaction and follow me—really appreciate it! 
PRs: https://github.com/prisma/prisma/pull/29365  https://github.com/prisma/prisma-engines/pull/5797 https://github.com/prisma/web/pull/7685

**Comment by lky141319** [comment]
Claiming bounty. 
Wallet: 0xd2268DEDca16b08aF4bb6Ad20cBa877D10cf8Ac6
GitHub: @lky141319

**Comment by lky141319** [comment]
Claiming bounty. 
Wallet: 0xd2268DEDca16b08aF4bb6Ad20cBa877D10cf8Ac6
GitHub: @lky141319

**Comment by lky141319** [comment]
Claiming bounty. 
Wallet: 0xd2268DEDca16b08aF4bb6Ad20cBa877D10cf8Ac6
GitHub: @lky141319

**Comment by karljv** [comment]
Please merge! PostGIS support would be great!

**Comment by jannis6023** [comment]
> Please merge! PostGIS support would be great!

Hi,
this docs page looks like it has already been merged in some way:
https://www.prisma.io/docs/orm/prisma-schema/postgresql-extensions

**Comment by lh0x00** [comment]
This feature includes multiple PRs, but only the document PRs have been merged; the implementations in https://github.com/prisma/prisma-engines/pull/5797 https://github.com/prisma/prisma/pull/29365 have not yet been merged.

**Comment by ussfranck** [comment]
> [@matthewmueller](https://github.com/matthewmueller) is planned before 2022?

Still not supported in 2026...

**Comment by lh0x00** [comment]
@ussfranck try https://www.npmjs.com/package/@vertexa/prisma, a fork version support postgis

**Comment by ofekBerko** [comment]
GTA 6 will be released before the feature becomes available

**Comment by MurzNN** [comment]
> GTA 6 will be released before the feature becomes available

Will GTA 6 support PostGIS? 😂

**Comment by ussfranck** [comment]
> [@ussfranck](https://github.com/ussfranck) try https://www.npmjs.com/package/@vertexa/prisma, a fork version support postgis

Oh Interesting...
Thank u very much 🙏🏽
